ekncontent/ekncontent/eknc-domain.c

```diff
@@ -739,40 +739,128 @@ static void
+  countdown->remaining = remaining;
+  countdown->callback = callback;
+  countdown->error_callback = error_callback;
+  countdown->data = data;
+
+  return countdown;
+}
+
+static void
+scatter_task_countdown_subtask_completed (ScatterTaskCountdown **countdown_ptr,
+                                          GError *error)
+{
+  ScatterTaskCountdown *countdown = *countdown_ptr;
+
+  /* If we received an error we should still continue to countdown, but
+   * allow the creator of the countdown to handle the error */
+  if (error && countdown->error_callback)
+    {
+      (*countdown->error_callback) (error, countdown->data);
+    }
 
+  if (--countdown->remaining == 0)
+    {
+      (*countdown->callback) (countdown->data);
+      g_free (countdown);
+      *countdown_ptr = NULL;
+    }
+}
+
+typedef struct {
+  /* Note that scatter_task_countdown_subtask_completed will free
+   * this member once the countdown has completed */
+  ScatterTaskCountdown *countdown;
+  EkncDomain *domain;
+  GPtrArray *models;
+  guint index;
+} ObjectResponseData;
+
+static ObjectResponseData *
+object_response_data_new (ScatterTaskCountdown *countdown,
+                          EkncDomain *domain,
+                          GPtrArray *models,
+                          guint index)
+{
+  ObjectResponseData *data = g_new0 (ObjectResponseData, 1);
+  data->countdown = countdown;
+  data->domain = domain;
+  data->models = models;
+  data->index = index;
+  return data;
+}
+
+static void
+on_received_all_model_objects (gpointer user_data)
+{
+  GTask *task = user_data;
+
+  /* We might have already been marked as completed earlier due to an
+   * error state. Don't return if that is the case */
   if (g_task_get_completed (task))
     return;
 
-  EkncContentObjectModel *model = eknc_domain_get_object_finish (domain, result, &error);
+  g_task_return_boolean (task, TRUE);
+  g_object_unref (task);
+}
+
+static void
+on_receive_model_error (GError *error, gpointer user_data)
+{
+  GTask *task = user_data;
+
   if (error != NULL)
     {
       g_task_return_error (task, error);
       return;
     }
+}
 
-  state->models = g_slist_append (state->models, model);
+static void
+on_object_response (GObject *source,
+                    GAsyncResult *result,
+                    gpointer user_data)
+{
+  g_autofree ObjectResponseData *object_response_data = user_data;
+  EkncDomain *domain = object_response_data->domain;
+  GError *error = NULL;
 
-  if (g_slist_length (state->models) == state->total_models)
-    {
-      g_task_return_boolean (task, TRUE);
-      return;
-    }
+  EkncContentObjectModel *model = eknc_domain_get_object_finish (domain, result, &error);
+
+  *(&g_ptr_array_index (object_response_data->models, object_response_data->index)) = model;
+  scatter_task_countdown_subtask_completed (&object_response_data->countdown, error);
 }
 
-void
+static void
+maybe_unref_object (gpointer data)
+{
+  if (data)
+    g_object_unref (data);
```

Comment:
I'm not sure I understand this point - is that because (if we don't care about ordering) we wouldn't have `NULL` slots for failed models, and thus we could set `g_object_unref` to be the finaliser func?